### PR TITLE
Changes for AZURE_CLIENT_SEND_CERTIFICATE_CHAIN

### PR DESCRIPTION
**Pull request by christothes**
This PR adds a new option to `internal/mock` which allows for remote addresses to be routed to the mock server. This change allows for more mock functional identity tests.

### CODE REVIEW

File: sdk/internal/mock/mock.go [hunk 1]
```diff
@@ -123,7 +127,20 @@ func (s *Server) Do(req *http.Request) (*http.Response, error) {
 		resp := s.getResponse()
 		return nil, resp.err
 	}
-	resp, err := s.srv.Client().Do(req)
+	var err error
+	var resp *http.Response
+	if s.routeAllRequestsToMockServer /*&& strings.Contains(req.URL.Host, "login")*/ {
```
Comment (jhendrixMSFT) [inline sdk/internal/mock/mock.go:132]: Nit, remove trailing comment.

File: sdk/internal/mock/mock.go [hunk 2]
```diff
@@ -123,7 +127,20 @@ func (s *Server) Do(req *http.Request) (*http.Response, error) {
...
+		originalURL := req.URL
+		mockUrl, _ := url.Parse(req.URL.String())
```
[the remark this answers is not in the document]
Comment (christothes) [reply sdk/internal/mock/mock.go:134]: OK, I just wasn't sure if that would be just copying the reference rather than creating a new instance with the values copied.

File: sdk/azidentity/azidentity_test.go [hunk 3]
```diff
@@ -27,6 +27,92 @@ const (
 	customHostString = "https://custommock.com/"
 )
 
+func getTenantDiscoveryResponse(url string) string {
+	return `{` +
```
Comment (chlowell) [inline sdk/azidentity/azidentity_test.go:31]: Backtick syntax allows multiline literals, so you don't have to concatenate every line

File: sdk/azidentity/environment_credential.go [hunk 4]
```diff
@@ -14,6 +14,10 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/internal/log"
 )
 
+const (
+	envVarSendCertChain = "AZURE_CLIENT_SEND_CERTIFICATE_CHAIN"
+)
```
Comment (chlowell) [inline sdk/azidentity/environment_credential.go:19]: Nitpicking:
```suggestion
const envVarSendCertChain = "AZURE_CLIENT_SEND_CERTIFICATE_CHAIN"
```

File: sdk/azidentity/environment_credential_test.go [hunk 5]
```diff
@@ -173,6 +179,50 @@ func TestEnvironmentCredential_UsernamePasswordSet(t *testing.T) {
...
+func TestEnvironmentCredential_UsernamePasswordSet_withCertificateChain_mock(t *testing.T) {
+	resetEnvironmentVarsForTest()
+
+	validateReq := func(req *http.Request) bool {
+		body, err := ioutil.ReadAll(req.Body)
+		if err == nil {
+			bodystr := string(body)
+			kvps := strings.Split(bodystr, "&")
+			assertion := strings.Split(kvps[0], "=")
+			token, _ := jwt.Parse(assertion[1], nil) //  func(token *jwt.Token) (interface{}, error) { return []byte(""), nil})
```
Comment (chlowell) [inline sdk/azidentity/environment_credential_test.go:191]: Vestigial comment?

File: sdk/azidentity/azidentity_test.go [hunk 6]
```diff
@@ -27,6 +27,92 @@ const (
 	customHostString = "https://custommock.com/"
 )
 
+func getTenantDiscoveryResponse(url string) string {
```
[the remark this answers is not in the document]
Comment (christothes) [reply sdk/azidentity/azidentity_test.go:30]: Initially I thought I would need to do that, but it wasn't required. I'll remove the param for now.

File: sdk/azidentity/client_certificate_credential_test.go [hunk 7]
```diff
@@ -82,6 +87,44 @@ func TestClientCertificateCredential_GetTokenSuccess_withCertificateChain(t *tes
...
 
+func TestClientCertificateCredential_GetTokenSuccess_withCertificateChain_mock(t *testing.T) {
+
+	validateReq := func(req *http.Request) bool {
+		body, err := ioutil.ReadAll(req.Body)
+		if err == nil {
+			bodystr := string(body)
+			kvps := strings.Split(bodystr, "&")
+			assertion := strings.Split(kvps[0], "=")
+			token, _ := jwt.Parse(assertion[1], nil)
```
Comment (seankane-msft) [inline sdk/azidentity/client_certificate_credential_test.go]: Does this line return an error as the second returned object? If so, it's probably best to check this error here

File: sdk/internal/mock/mock.go [hunk 8]
```diff
@@ -123,7 +127,20 @@ func (s *Server) Do(req *http.Request) (*http.Response, error) {
 		resp := s.getResponse()
 		return nil, resp.err
 	}
-	resp, err := s.srv.Client().Do(req)
+	var err error
+	var resp *http.Response
+	if s.routeAllRequestsToMockServer {
+		originalURL := req.URL
+		mockUrl := *req.URL
+		srvUrl, _ := url.Parse(s.srv.URL)
```
Comment (seankane-msft) [inline sdk/internal/mock/mock.go:135]: Unlikely, but if there is an error here we should return the error and an empty `&http.Response{}`

File: sdk/azidentity/client_certificate_credential_test.go (hunk 7 above)
Comment (christothes) [reply sdk/azidentity/client_certificate_credential_test.go]: yes, but in the end the test will fail either way, I believe.

File: sdk/internal/mock/mock.go (hunk 8 above)
Comment (christothes) [reply sdk/internal/mock/mock.go:135]: I'm not sure it is worth handling this extreme edge case.

File: sdk/azidentity/client_certificate_credential_test.go (hunk 7 above)
Comment (seankane-msft) [reply sdk/azidentity/client_certificate_credential_test.go]: It will, but if there is an error on parsing, it will fail because of a nil pointer exception whereas this will help with diagnosability

File: sdk/internal/mock/mock.go (hunk 8 above)
Comment (chlowell) [reply sdk/internal/mock/mock.go:135]: Handling unlikely errors is the quintessence of Go programming 🤣 

But shouldn't we return `nil` instead of `&http.Response{}`?

File: sdk/azidentity/azidentity_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (chlowell) [reply sdk/azidentity/azidentity_test.go]: Might as well be a constant then?

File: sdk/internal/mock/mock.go (hunk 8 above)
Comment (christothes) [reply sdk/internal/mock/mock.go:135]: Fair enough 😄

File: sdk/azidentity/client_certificate_credential_test.go (hunk 7 above)
Comment (christothes) [reply sdk/azidentity/client_certificate_credential_test.go]: actually, in this case we return an error even though the jwt parse is successful because we are passing nil for the validation function. I think we could check for JWT being null here instead though.

File: sdk/azidentity/azidentity_test.go [hunk 9]
```diff
@@ -23,10 +28,114 @@ const (
...
+		":method": "POST",
+		":path": "/fake-tenant/oauth2/v2.0/token",
+		":scheme": "https",
+		"Accept-Encoding": "gzip",
+		"Content-Length": "2",
+		"Content-Type": "application/x-www-form-urlencoded; charset=utf-8",
+		"User-Agent": "azsdk-go-azidentity/v0.12.1 azsdk-go-azcore/v0.21.0 (go1.16.7; linux)"
+		},
+		"RequestBody": {},
+		"StatusCode": 400,`
```
Comment (chlowell) [inline sdk/azidentity/azidentity_test.go:115]: Got the next request from the recording dangling here
```suggestion
```

File: sdk/azidentity/azidentity_test.go [hunk 10]
```diff
@@ -23,10 +28,114 @@ const (
...
 )
 
+func validateJWTRequestContainsHeader(t *testing.T, headerName string) mock.ResponsePredicate {
+	return func(req *http.Request) bool {
+		body, err := ioutil.ReadAll(req.Body)
+		if err == nil {
```
Comment (chlowell) [inline sdk/azidentity/azidentity_test.go:121]: Nitpicking: `err != nil` would be more idiomatic (handler immediately follows the call that returned the error; happy path less indented)

File: sdk/azidentity/azidentity_test.go [hunk 11]
```diff
@@ -23,10 +28,114 @@ const (
 
 // constants for this file
 const (
-	envHostString    = "https://mock.com/"
-	customHostString = "https://custommock.com/"
+	envHostString              = "https://mock.com/"
+	customHostString           = "https://custommock.com/"
+	getTenantDiscoveryResponse = `{
```
Comment (chlowell) [inline sdk/azidentity/azidentity_test.go:33]: I'd probably put this and the validation function in `client_certificate_credential_test.go`. This file's more about high-level tests and broadly useful helpers
```suggestion
	tenantDiscoveryResponse = `{
```

**Review by chlowell — APPROVED** [review]
Thanks for doing this! I hope you enjoyed writing some Go :hamster:
